common/src/tap/checks/value_check.rs

```diff
@@ -0,0 +1,271 @@
+        let query_ids_clone = query_ids.clone();
+        let model_handle = tokio::spawn(async move {
+            loop {
+                let model = rx_cost_model.recv().await;
+                match model {
+                    Some(value) => {
+                        let deployment_id = value.deployment_id;
+
+                        if let Some(query) = cache.lock().unwrap().get_mut(&deployment_id) {
+                            let _ = query.insert_model(value).inspect_err(|err| {
+                                tracing::error!(
+                                    "Error while compiling cost model for deployment id {}. Error: {}",
+                                    deployment_id, err
+                                )
+                            });
+                        } else {
+                            match CostModelCache::new(value) {
+                                Ok(value) => {
+                                    cache.lock().unwrap().insert(deployment_id, value);
+                                }
+                                Err(err) => {
+                                    tracing::error!(
+                                        "Error while compiling cost model for deployment id {}. Error: {}",
+                                        deployment_id, err
+                                    )
+                                }
+                            }
+                        }
+                    }
+                    None => break,
+                }
+            }
+        });
+
+        // we use two different handles because in case one channel breaks we still have the other
+        let query_handle = tokio::spawn(async move {
+            loop {
+                let query = rx_query.recv().await;
+                match query {
+                    Some(query) => {
+                        query_ids_clone
+                            .lock()
+                            .unwrap()
+                            .insert(query.signature, query);
+                    }
+                    None => break,
+                }
+            }
+        });
+
+        Self {
+            cost_model_cache,
+            model_handle,
+            query_ids,
+            query_handle,
+        }
+    }
+}
+
+impl Drop for MinimumValue {
+    fn drop(&mut self) {
+        self.model_handle.abort();
+        self.query_handle.abort();
+    }
+}
+
+impl MinimumValue {
+    fn get_agora_query(&self, query_id: &Signature) -> Option<AgoraQuery> {
+        self.query_ids.lock().unwrap().remove(query_id)
+    }
+
+    fn get_expected_value(&self, query_id: &Signature) -> anyhow::Result<u128> {
+        // get query from key
+        let agora_query = self
+            .get_agora_query(query_id)
+            .ok_or(anyhow!("No query found"))?;
+
+        // get agora model for the allocation_id
+        let mut cache = self.cost_model_cache.lock().unwrap();
+        // on average, we'll have zero or one model
```

Comment:
I'd expect that we can get more than 1 model in the case the model was recently updated, and we are still in the period where expect the query to be priced according to a previous model or the new one.

Reply:
indeed, we'll have more than one model when recently updated. But I expect that this is not that often. That's the reason for my comment. On average we'll have one model, or even zero if it's not set.